certora/specs/ERC20.spec

```diff
@@ -0,0 +1,256 @@
+│ Invariant: balance of address(0) is 0                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant zeroAddressNoBalance()
+    balanceOf(0) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only the token holder (or a permit) can increase allowance. The spender can decrease it by using it          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyHolderOrSpenderCanChangeAllowance(env e, method f){
+    requireInvariant totalSupplyIsSumOfBalances();
+
+    calldataarg args;
+    address holder;
+    address spender;
+
+    uint256 allowanceBefore = allowance(holder, spender);
+    f(e, args);
+    uint256 allowanceAfter = allowance(holder, spender);
+
+    assert (
+        allowanceAfter > allowanceBefore
+    ) => (
+        (f.selector == sig:approve(address,uint256).selector           && e.msg.sender == holder) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+
+    assert (
+        allowanceAfter < allowanceBefore
+    ) => (
+        (f.selector == sig:transferFrom(address,address,uint256).selector && e.msg.sender == spender) ||
+        (f.selector == sig:approve(address,uint256).selector              && e.msg.sender == holder ) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: transfer behavior and side effects                                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule transfer(env e) {
+    requireInvariant totalSupplyIsSumOfBalances();
+    require nonpayable(e);
+
+    address holder = e.msg.sender;
+    address recipient;
+    address other;
+    uint256 amount;
+
+    // cache state
+    uint256 holderBalanceBefore    = balanceOf(holder);
+    uint256 holderVotingPowerBefore = delegatedVotingPower(delegatee(holder));
+    uint256 recipientBalanceBefore = balanceOf(recipient);
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(recipient));
+    uint256 otherBalanceBefore     = balanceOf(other);
+
+    // run transaction
+    transfer@withrevert(e, recipient, amount);
+
+    // check outcome
+    if (lastReverted) {
+        assert holder == 0 || recipient == 0 || amount > holderBalanceBefore ||
+            // Handle overflows in delegation, should not be possible.
+            recipientVotingPowerBefore + amount > max_uint256 || holderVotingPowerBefore < amount ;
```

Comment:
we prove that lastReverted => these, but we miss the <= (or !these => !lastReverted, in the else)

Reply:
maybe it's better to just write lastReverted <=> conditions, btw (and check the other things separately)

Reply:
actually, the other branch of the checks `<=` part, or is it not what you mean?

Reply:
The other branch could in theory check the other direction, but I don't think it does it consistently. I think what @MathisGD suggested is safer if we can do it. For example here, if it was removed in the code that it reverts when `holder == 0`, then this specification would still go through

Reply:
This would be a reason to deviate from OZ specification btw

Reply:
The caveat of this specification is that liveness and safety are mixed up together. This makes it hard to figure out what is happening and why. 

However, how do we want to handle this? If I understand correctly the above comments, one suggestion is to refine/strengthen the else-branch of this specification.

Suppose now that we have two sort of specs for this specific case. One that checks that _reverts_ will be triggered when it's expected and one that checks that balances are properly updated. Let's refer to the former as transfer-revert rules and the latter transfer-ok rules for sake of clarity. 

We have two choices, either we strenghten the else-branch of the conditional in this spec to explicitly  ensure all conditions like `assert holder != 0` or, split the proof in a transfer-ok (without `@withrevert`) and several transfer-revert rules. For instance, a transfer-revert for `holder == 0 => lastReverted`.

I think it would be better to add explicit rules for revert-conditions. This could make the transfer-ok rules proofs easier to read and maintain (less explicit assertions) and revert points easier to spot (explicit point of revertion checked by a dedicated rule yield more accurate counter examples).

Interestingly, in this setting (Morpho token) we can add revert checks in a separate spec that doesn't interfere with this spec.